File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/PersonalTabDataCache.kt
package com.example.momentra.ui.shell.personal

import com.example.momentra.data.api.ActivityItemDto
import com.example.momentra.data.api.PersonalPulseDto
import java.util.concurrent.ConcurrentHashMap

/** In-memory SWR cache for Personal tab datasets (pulse + activity preview). */
object PersonalTabDataCache {
    data class Entry(
        val pulse: PersonalPulseDto,
        val activities: List<ActivityItemDto>,
        val savedAtMs: Long = System.currentTimeMillis(),
    )

    private val store = ConcurrentHashMap<String, Entry>()

    fun cacheKey(momentId: String?): String = momentId?.takeIf { it.isNotBlank() } ?: "__personal_default__"

    fun peek(momentId: String?): Entry? = store[cacheKey(momentId)]

    fun put(momentId: String?, pulse: PersonalPulseDto, activities: List<ActivityItemDto>) {
        store[cacheKey(momentId)] = Entry(pulse, activities)
    }

    fun clear() {
        store.clear()
    }
}

File: momentra/momentra/Shell/Personal/Shared/PersonalTabDataCache.swift
import Foundation

/// In-memory SWR cache for Personal tab datasets (pulse + activity preview).
enum PersonalTabDataCache {
    struct PulseEntry {
        let pulse: APIClient.PersonalPulsePayload
        let activities: [APIClient.ActivityItemPayload]
        let savedAtMs: TimeInterval
    }

    private static var pulseStore: [String: PulseEntry] = [:]
    private static var lifePayload: APIClient.PersonalLifePayload?

    static func cacheKey(momentId: String?) -> String {
        guard let momentId, !momentId.isEmpty else { return "__personal_default__" }
        return momentId
    }

    static func peekPulse(momentId: String?) -> PulseEntry? {
        pulseStore[cacheKey(momentId: momentId)]
    }

    static func putPulse(momentId: String?, pulse: APIClient.PersonalPulsePayload, activities: [APIClient.ActivityItemPayload]) {
        pulseStore[cacheKey(momentId: momentId)] = PulseEntry(
            pulse: pulse,
            activities: activities,
            savedAtMs: Date().timeIntervalSince1970 * 1000
        )
    }

    static func peekLife() -> APIClient.PersonalLifePayload? {
        lifePayload
    }

    static func putLife(_ payload: APIClient.PersonalLifePayload) {
        lifePayload = payload
    }

    static func clear() {
        pulseStore.removeAll()
        lifePayload = nil
    }
}

struct PersonalPulseTabData {
    let pulse: APIClient.PersonalPulsePayload
    let activities: [APIClient.ActivityItemPayload]
}
